models/contracts/optimism/contracts_optimism_contract_mapping.sql

```diff
@@ -0,0 +1,315 @@
+        else b.creator_address
+      end as contract_factory
+      {% endif %}
+      ,b.contract_address
+      ,b.created_time
+      ,b.creation_tx_hash
+      ,b.trace_element
+      ,b.is_self_destruct
+    {% if loop.first -%}
+    from base_level as b
+    left join base_level as u
+      on b.creator_address = u.contract_address
+    {% else -%}
+    from level{{i-1}} as b
+    left join base_level as u
+      on b.creator_address = u.contract_address
+    {% endif %}
+)
+{%- endfor %}
+
+,creator_contracts as (
+  select 
+    f.creator_address
+    ,f.contract_factory
+    ,f.contract_address
+    ,coalesce(cc.contract_project, ccf.contract_project) as contract_project 
+    ,f.created_time
+    -- check if the contract is an immediate self-destruct contract
+    ,coalesce(f.is_self_destruct, 
+      case 
+        when exists (
+          select 1 
+          from {{ source('optimism', 'traces') }} as sd 
+          where 
+            f.creation_tx_hash = sd.tx_hash
+            and f.trace_element = sd.trace_address[1]
+            and sd.type = 'suicide'
+            {% if is_incremental() %} -- this filter will only be applied on an incremental run 
+            and sd.block_time >= date_trunc('day', now() - interval '1 week')
+            {% endif %}
+        ) then true 
+      else false 
+    end) as is_self_destruct
+    ,f.creation_tx_hash
+    ,f.trace_element 
+  from level{{max_levels}} as f
+  left join {{ ref('contracts_optimism_contract_creator_address_list') }} as cc 
+    on f.creator_address = cc.creator_address
+  left join {{ ref('contracts_optimism_contract_creator_address_list') }} as ccf
+    on f.contract_factory = ccf.creator_address
+  where f.contract_address is not null
+ )
+,combine as (
+  select 
+    cc.creator_address
+    ,cc.contract_factory
+    ,cc.contract_address
+    ,coalesce(cc.contract_project, oc.namespace) as contract_project 
+    ,oc.name as contract_name 
+    ,cc.created_time
+    ,coalesce(cc.is_self_destruct, false) as is_self_destruct
+    ,'creator contracts' as source
+    ,cc.creation_tx_hash
+    ,cc.trace_element
+  from creator_contracts as cc 
+  left join {{ source('optimism', 'contracts') }} as oc 
+    on cc.contract_address = oc.address 
+
+  union all 
+
+  -- CHUXIN: is this part necessary?
+  select 
+    NULL as creator_address
+    ,NULL as contract_factory
+    ,oc.address as contract_address
+    ,oc.namespace as contract_project
+    ,oc.name as contract_name
+    ,oc.created_at as created_time
+    ,coalesce(cc.is_self_destruct, false) as is_self_destruct
+    ,'decoded contracts' as source
+    ,cc.creation_tx_hash
+    ,cc.trace_element
+  from {{ source('optimism', 'contracts') }} as oc 
+  join creator_contracts as cc 
+    on oc.address = cc.contract_address
+  where 
+    true
+    {% if is_incremental() %} -- this filter will only be applied on an incremental run 
+    and created_at >= date_trunc('day', now() - interval '1 week')
+    {% endif %}
```

Comment:
do you think this part is doing the same as line 160 - 173? 
since it's a join so it's not adding anything addition from `optimism.contracts` @MSilb7

Reply:
slightly different.

first one is saying "take all mapped contracts," second one is saying "take all dune decoded contracts."

But since we're going through all contracts now, yes it looks like we can just have 160-173.